### PR TITLE
Add shapes tutorial

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Tutorial for https://github.com/pytorch/pytorch/issues/97118

### CODE REVIEW

File: recipes_source/recipes/reasoning_about_shapes.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
+"""
+Reasoning about Shapes in PyTorch
+=================================
+
+When writing models with PyTorch, it is commonly the case that the parameters
+to a given layer depend on the shape of the output of the previous layer. For
+example, the ``in_features`` of an ``nn.Linear`` layer are dependent on
+``.size(-1)`` of the input. For some layers, the shape computation involves
+complex equations (e.g. convolution operations).
```
Comment (svekars) [inline recipes_source/recipes/reasoning_about_shapes.py:9]:
```suggestion
complex equations, for example convolution operations.
```

File: recipes_source/recipes/reasoning_about_shapes.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+start = timeit.default_timer()
+t_large = torch.rand(2**10, 3, 2**16, 2**16, device="meta")
+end = timeit.default_timer()
+
+print(conv(t_large))
+print(f"Time taken: {end-start}")
+
+#####################################################
+# Consider an aribtrary network such as the following
```
Comment (svekars) [inline recipes_source/recipes/reasoning_about_shapes.py:42]:
```suggestion
# Consider an aribtrary network such as the following:
```

File: recipes_source/recipes/reasoning_about_shapes.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
+"""
+Reasoning about Shapes in PyTorch
+=================================
+
+When writing models with PyTorch, it is commonly the case that the parameters
+to a given layer depend on the shape of the output of the previous layer. For
+example, the ``in_features`` of an ``nn.Linear`` layer are dependent on
+``.size(-1)`` of the input. For some layers, the shape computation involves
```
Comment (jbschlosser) [inline recipes_source/recipes/reasoning_about_shapes.py:8]:
```suggestion
example, the ``in_features`` of an ``nn.Linear`` layer must match the
``size(-1)`` of the input. For some layers, the shape computation involves
```

File: recipes_source/recipes/reasoning_about_shapes.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        x = self.pool(F.relu(self.conv2(x)))
+        x = torch.flatten(x, 1) # flatten all dimensions except batch
+        x = F.relu(self.fc1(x))
+        x = F.relu(self.fc2(x))
+        x = self.fc3(x)
+        return x
+
+
+with torch.device("meta"):
+  net = Net()
```
Comment (jbschlosser) [inline recipes_source/recipes/reasoning_about_shapes.py:69]:
```suggestion
    net = Net()
```
(to match previous indentation)

File: recipes_source/recipes/reasoning_about_shapes.py [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        x = self.pool(F.relu(self.conv1(x)))
+        x = self.pool(F.relu(self.conv2(x)))
+        x = torch.flatten(x, 1) # flatten all dimensions except batch
+        x = F.relu(self.fc1(x))
+        x = F.relu(self.fc2(x))
+        x = self.fc3(x)
+        return x
+
+
+with torch.device("meta"):
```
Comment (jbschlosser) [inline recipes_source/recipes/reasoning_about_shapes.py:68]: nit: I suggest moving this network initialization to the next section and leave this code snippet as simply the network definition

File: recipes_source/recipes/reasoning_about_shapes.py [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+  net = Net()
+
+#############################################################################
+# We can extend this to an entire network if we wish to reason about the shapes
+# of each layer by registering a forward hook to each layer that prints the
+# shape of the output.
```
Comment (jbschlosser) [inline recipes_source/recipes/reasoning_about_shapes.py:74]:
```suggestion
#############################################################################
# We can view the intermediate shapes within an entire network by registering a
# forward hook to each layer that prints the shape of the output.
```

File: recipes_source/recipes/reasoning_about_shapes.py [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+def fw_hook(module, input, output):
+  print(f"Shape of output to {module} is {output.shape}.")
```
Comment (jbschlosser) [inline recipes_source/recipes/reasoning_about_shapes.py:77]:
```suggestion
  print(f"Shape of output for {module} is {output.shape}.")
```

(also 4 space indentation to match previous code snippets)

File: recipes_source/recipes/reasoning_about_shapes.py [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+for name, layer in net._modules.items():
+  layer.register_forward_hook(fw_hook)
```
Comment (jbschlosser) [inline recipes_source/recipes/reasoning_about_shapes.py:80]:
```suggestion
    layer.register_forward_hook(fw_hook)
```

File: recipes_source/recipes/reasoning_about_shapes.py (hunk 5 above)
Comment (jbschlosser) [reply recipes_source/recipes/reasoning_about_shapes.py:68]: Also I suggest adding a short comment explaining that any tensors created underneath the context manager will be on the 'meta' device. You could also create the input underneath within the context manager

**Review by jbschlosser — APPROVED** [review]
Thanks for adding this! It reads really well IMO.

Just a couple formatting / flow nits but otherwise solid :)

File: recipes_source/recipes/reasoning_about_shapes.py [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+#############################################################################
+# We can extend this to an entire network if we wish to reason about the shapes
+# of each layer by registering a forward hook to each layer that prints the
+# shape of the output.
+
+def fw_hook(module, input, output):
+  print(f"Shape of output to {module} is {output.shape}.")
+
+for name, layer in net._modules.items():
```
Comment (albanD) [inline recipes_source/recipes/reasoning_about_shapes.py:79]: Never access private attributes ;) 
Also if you want nicer prints, you can use `net.named_modules()` to get nice names to print in the hook.

**Review by albanD — COMMENTED** [review]
Sounds good! Only small comment!

File: recipes_source/recipes/reasoning_about_shapes.py [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+=================================
+
+When writing models with PyTorch, it is commonly the case that the parameters
+to a given layer depend on the shape of the output of the previous layer. For
+example, the ``in_features`` of an ``nn.Linear`` layer must match the
+``size(-1)`` of the input. For some layers, the shape computation involves
+complex equations, for example convolution operations.
+
+One way around this is to run the forward pass with random inputs, but this is
+wasteful in terms of compute.
```
Comment (albanD) [inline recipes_source/recipes/reasoning_about_shapes.py:12]: nit: memory and compute

**Review by albanD — APPROVED** [review]
Sounds good!

File: _static/img/thumbnails/cropped/reasoning-about-shapes.PNG
[the remark this answers is not in the document]
Comment (svekars) [reply _static/img/thumbnails/cropped/reasoning-about-shapes.PNG:1]: You can just use the generic image: https://github.com/pytorch/tutorials/blob/main/_static/img/thumbnails/cropped/generic-pytorch-logo.png

**Review by svekars — APPROVED** [review]

**Pull request merged**